data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/S3SinkWorker.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.plugins.sink;
+
+import java.io.IOException;
+import java.util.NavigableSet;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.TimeUnit;
+import org.apache.commons.lang3.time.StopWatch;
+import org.mapdb.DB;
+import org.mapdb.DBMaker;
+import org.mapdb.Serializer;
+import org.opensearch.dataprepper.model.event.Event;
+import org.opensearch.dataprepper.model.types.ByteCount;
+import org.opensearch.dataprepper.plugins.sink.accumulator.BufferType;
+import org.opensearch.dataprepper.plugins.sink.accumulator.BufferTypeOptions;
+import org.opensearch.dataprepper.plugins.sink.accumulator.InMemoryBuffer;
+import org.opensearch.dataprepper.plugins.sink.accumulator.LocalFileBuffer;
+import org.opensearch.dataprepper.plugins.sink.codec.Codec;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import software.amazon.awssdk.services.s3.S3Client;
+
+/**
+ * Class responsible for threshold check and instantiation of {@link BufferType}
+ */
+public class S3SinkWorker {
+
+    private static final Logger LOG = LoggerFactory.getLogger(S3SinkWorker.class);
+    private final S3Client s3Client;
+    private final S3SinkConfig s3SinkConfig;
+    private final Codec codec;
+    private final int numEvents;
+    private final ByteCount byteCapacity;
+    private final long duration;
+
+    /**
+     * @param s3Client
+     * @param s3SinkConfig
+     * @param codec
+     */
+    public S3SinkWorker(final S3Client s3Client, final S3SinkConfig s3SinkConfig, final Codec codec) {
+        this.s3Client = s3Client;
+        this.s3SinkConfig = s3SinkConfig;
+        this.codec = codec;
+        numEvents = s3SinkConfig.getThresholdOptions().getEventCount();
+        byteCapacity = s3SinkConfig.getThresholdOptions().getMaximumSize();
+        duration = s3SinkConfig.getThresholdOptions().getEventCollect().getSeconds();
+    }
+
+    /**
+     * Accumulates data from buffer and store in local-file/in-memory.
+     *
+     * @param eventQueue
+     */
+    public void bufferAccumulator(BlockingQueue<Event> eventQueue) {
+        boolean isFileUploadedToS3 = Boolean.FALSE;
+        DB eventDb = null;
+        NavigableSet<String> bufferedEventSet = null;

```

Comment:
Why is this a `NavigableSet`? I'm not even sure you need a collection.

I suggested elsewhere that you pass in a `List<Event>` to this method. Just provide that `List` to the accumulators. Those accumulators can then iterate over the lists and add each `Event` into their respective buffers.

Reply:
These changes will be addressed as incremental changes, and they will be available in a separate PR.